metagov/metagov/core/identity.py

```diff
@@ -0,0 +1,239 @@
+from contextlib import suppress
+from enum import Enum
+
+import json, random
+from django.db import models, IntegrityError
+
+from metagov.core.models import Community
+
+
+class MetagovID(models.Model):
+    """Metagov ID table links all public_ids to a single internal representation of a user. When data
+    associated with public_ids conflicts, primary_ID is used.
+
+    Fields:
+
+    community: foreign key - metagov community the user is part of
+    internal_id: integer - unique, secret ID
+    public_ids: list of integers - public IDs shared with Driver, callers, etc. must have at least one item
+    primary_id: integer - used to resolve conflicts if there are multiple public_ids associated with a user
+
+    FIXME: I don't like that linked_accounts links to this model as metagov_id but it's actually
+    linking to one of the public_ids (which, admittedly should resolve to a single metagov_id)"""
+
+    community = models.ForeignKey(Community)
+    internal_id = models.PositiveIntegerField(unique=True)
+    public_ids = models.JSONField(default=list)
+    primary_id = models.PositiveIntegerField(unique=True)
+
+    # Internal logic
+
+    def add_id(self, new_id):
+        if not self.public_ids:
+            self.primary_id = new_id
+        self.public_ids.append(new_id)
+
+    def change_primary_id(self, new_primary_id):
+        if new_primary_id not in self.public_ids:
+            raise ValueError(f"New primary_id {new_primary_id} not in public IDs")
+        self.primary_id = new_primary_id
+
+
+class LinkType(Enum):
+    OAUTH = "oauth"
+    MANUAL_ADMIN = "manual admin"
+    EMAIL_MATCHING = "email matching"
+    UNKNOWN = "unknown"
+
+
+class LinkQuality(Enum):
+    STRONG_CONFIRM = "confirmed (strong)"
+    WEAK_CONFIRM = "confirmed (weak)"
+    UNCONFIRMED = "unconfirmed"
+    UNKNOWN = "unknown"
+
+
+class LinkedAccounts(models.Model):
+    """Contains information about specific platform account linked to user
+
+    Fields:
+
+    metagov_id: integer - must match ID in public_ids col of a row in metagov_IDs table
+    community: foreign key - metagov community the user is part of
+    community_platform_id: string (optional) - distinguishes between ie two Slacks in the same community
+    platform_type: string - ie Github, Slack
+    platform_identifier: string - ID, username, etc, unique to the platform (or unique to community_platform_id)
+    custom_data: dict- optional additional data for linked platform account
+    link_type: string (choice) - method through which account was linked
+    link_quality: string (choice) - metagov's assessment of the quality of the link (depends on method)
+   """
+
+    metagov_id = models.PositiveBigIntegerField()
```

Comment:
Why use an integer here instead of `ForeignKey(MetagovID)`?

Reply:
Let's say I, Shauna, have signed up to Metagov and authenticated my Slack account, creating the following instances:

`MetagovID: pk 1, internal_id: 121212, public_ids: [343434], primary_id: 343434`
`LinkedAccount: metagov_id: 343434, platform_type: Slack, platform_identifier: shaunagm`

The community adds a Github repository and all of the contributors to that repository get identities associated with them, but they aren't connected to existing identities. So we get the following additional models:

`MetagovID: pk 2, internal_id: 565656, public_ids: [787878], primary_id: 787878`
`LinkedAccount: metagov_id: 787878, platform_type: Github, platform_identifier: shauna_gm`

Now you as an administrator says, oh, hey, those are the same person! Let me merge them. This creates the following:

`MetagovID: pk 1, internal_id: 121212, public_ids: [343434, 787878], primary_id: 343434`
`LinkedAccount: metagov_id: 343434, platform_type: Slack, platform_identifier: shaunagm`
`LinkedAccount: metagov_id: 787878, platform_type: Github, platform_identifier: shauna_gm`

Meanwhile, the second MetagovID is deleted. However, we don't need to make any other change except adding to the public_ids of the first MetagovID, because the LinkedAccounts aren't foreign-keyed to the MetagovID object. Their metagov_ids are still valid, they've just been transferred to a different MetagovID instance.

This is the data model I thought we'd consensed on but I am open to changing it. It would make many things easier if we could just foreign key to MetagovID -- in this PR I mistakenly used code like `user.linked_accounts.filter` which assumes a foreignkey relationship, and actually I need to go in and manually get the linked_accounts for the user if we're keeping the structure above. But I don't know if that's so much work that we want to re-think how we're doing merging.

Reply:
nevermind, I see. The metagov ID could be one of the IDs in the jsonfield array.

Reply:
Here's a slightly different approach to consider:

Changing `LinkedAccounts.metagov_id` to be a `ForeignKey(MetagovID)`, and changing the `MetagovID.public_ids` field from a JSONField to a `linked_ids = models.ManyToManyField('self', symmetrical=False)
` ([about many to many self](https://stackoverflow.com/questions/6218175/how-to-implement-followers-following-in-django)). 
Then we'd have a MetagovID table with a bunch of rows that are actually linked together (indicating that a merge has occurred). In the merge function, instead of *deleting* a row from MetagovID, we would just add the link `primary_instance.linked_ids.add(secondary_instance)`. We wouldn't need to modify LinkedAccounts when a merge happens, because that ForeignKey is still linking to a real record.


I don't feel strongly about either way. I think I have a slight preference to using Django relations instead of maintaining our own array of `public_ids`. IIUC that array is essentially a list of IDs from _deleted rows_ of the MetagovID table. The database constrains that `primary_id` is unique, but once we delete rows is it fair game for the database to create new records with the same key? For example if I have accounts 1 and 2, I perform a "merge" which deletes record 2 and modifies record 1 to look like `primary: 1, public_ids: [2]`. THEN I create an entirely new unrelated account, which may be assigned metagov_id 2 (?) (satisfies uniqueness constraint b/c we deleted the old record).

Reply:
`is it fair game for the database to create new records with the same key? For example if I have accounts 1 and 2, I perform a "merge" which deletes record 2 and modifies record 1 to look like primary: 1, public_ids: [2]. THEN I create an entirely new unrelated account, which may be assigned metagov_id 2 (?) (satisfies uniqueness constraint b/c we deleted the old record).` - with the constraints we have now, yes, this could happen, the database wouldn't stop it. We could add another uniqueness check on model save to stop it, but it's worth investigating if a different model wouldn't be simpler and better.

In your proposal above, when we link two accounts together, how do we know which is primary? I guess we could keep the separate primary field and make it boolean, although that wouldn't prevent accidentally labeling two linked metagovID instances with primary=True.

Reply:
> that wouldn't prevent accidentally labeling two linked metagovID instances with primary=True.

We could validate it in `save()`

Reply:
> This is the data model I thought we'd consensed on but I am open to changing it. It would make many things easier if we could just foreign key to MetagovID -- in this PR I mistakenly used code like user.linked_accounts.filter which assumes a foreignkey relationship, and actually I need to go in and manually get the linked_accounts for the user if we're keeping the structure above. But I don't know if that's so much work that we want to re-think how we're doing merging.

Yeah. They are really similar. That's true, the search/filter capability might be easier to implement if we use the database relations. I would _guess_ it will be faster, too, to do lookups on a column vs lookups for an item in a JSONField array. I'm not sure about that though,  postgres might do some fancy indexing.

Either approach is fine for me, your call!